### PR TITLE
Add image rotation for detecting more faces and dealing with awkward angles

### PR DESCRIPTION
**Pull request by torzdf**
This change allows for the detection of more faces, and often a more satisfactory swap. Dlib has problems recognizing faces tilted over more than 90 degrees from center, so this helps to mitigate that with an optional flag at the extract stage (-r --rotate_images).

If activated, then at extract stage, if a face is not detected the image is rotated 90 degrees and another attempt is made. This pattern is repeated, so in total an image will be scanned at: normal, 90 degrees, 180 degrees, 270 degrees.

If an image is found, then the rotation stops, and the process continues. Obviously, with lots of frames without faces or with faces at awkward angles, this WILL slow down the process, hence why it is an optional flag.

In order to support this, a new item is entered into alignments.json that stores the rotation of the image. At convert stage, this is read back from the file and the correct rotation is done on the output frame to apply the faceswap, before rotating the frame back to the correct orientation.

I needed to amend 5 files for this, but the changes are fairly cosmetic, and *shouldn't* break anything else.

Please let me know thoughts, and merge if possible!

Any questions, let me know.

Cheers

### CODE REVIEW

**Comment by torzdf** [comment]
Oh, and I found a bug during testing, hence my revert. You should only focus on the final commit

**Comment by iperov** [comment]
have you example where your fix can detect more faces without flip-glitches by aligner?

**Comment by Clorr** [comment]
@torzdf No problem for the revert, it will be squashed.

**Comment by torzdf** [comment]
@iperov? How do you mean? I have nothing I can share, no, but results have been fine for me. Please feel free to test, any feedback is good.

This doesn't do anything complicated other than rotate the image. The alignments pulled from dlib are correct for the rotated image. The image is then rotated again (by the same amount) for applying the swap, so there shouldn't be any glitches here? I haven't done anything complicated like transposing dlib matrices. That would probably be a lot more efficient, but was beyond me at this stage.

**Comment by Clorr** [comment]
Related to #197

**Comment by torzdf** [comment]
Oh and I removed the count variable from handleImage because it wasn't doing anything.

**Comment by babilio** [comment]
This is working well and #259 that I created it works pretty fast on extraction 8,000 images ~ 40mins and about ~1500 where sideways or upside down and found with this. However the convert seems to be taking much longer. I have a local branch with this and #260 and it is taking 2hrs for convert, I am not sure if it is because of this pull request particularly. I can test more tomorrow. Thanks for creating this!

**Comment by torzdf** [comment]
Hmmm, I did not notice any difference on convert. I found extract was slower if there were lots of images with no faces/at awkward angles, but that is expected.

Convert has to do less work: it checks the rotation from alignments.json, (which it already has loaded) and only rotates the image if a non zero value is found. Obviously if a non zero value is found, it needs to rotate the image to perform the swap and then rotate back to the correct orientation, but the speed loss should be marginal.

If there isn't an alignments.json, it should just default to 0 rotation, so would never rotate the image.

Would be interested if anyone else is having issues, I will test more, but am unable to replicate at the moment.

**Comment by torzdf** [comment]
It isn't automatic. It's run with the command line switch -r (--rotate-images), so is fully optional.

If the switch isn't specified, then no rotation will be attempted, and it will run as it always has

**Comment by torzdf** [comment]
Also indirectly relates to issue: #224 where this was discussed

**Comment by bryanlyon** [comment]
I apologize, my message was incomplete.  I mean rotate explicitly when told to.  As in don't try to check for faces upright first, but when given the -r command only try with rotation.  Since trying upright first may find the same (incorrect) face and then skip the rotate.  Perhaps even with explicit angles listed (-r 90, -r 180, or 270).

**Comment by torzdf** [comment]
Ok, I see what you're saying. That should be a relatively straightforward add, but I think I would leave it as an improvement for after it gets merged, if it does, in its current form.

I may be wrong, but I envisage that most people will activate it when they know they have faces in their target video that may be problematic, so it makes sense to pass all frames in one go, rather than doing upright first and then doing a second pass to attempt to get rotated faces (it would be slower that way).

That said, adding options to the command line argument to check normal and rotated or just rotated may be something I could look to instigate down the line, but I don't see it as a must have. Ultimately it will only perform rotation on frames that it can't find a face on, so if a video is passed in with 99% upright faces, the slowdown will not be high.

**Comment by bryanlyon** [comment]
In my experience, videos with tilted faces will have both an upright face and a tilted face.  The upright face is often not the face that is desired.  This PR solves the problem only if there is not upright face.

My main concern is that if this is added as it is, then future modification to enable that flexibility would harm backward compatibility with command line options.  Obviously that isn't a huge deal, but changing the function of a command line switch is something we should avoid.

**Comment by bryanlyon** [comment]
Apologies for the double post, but I came up with a way to make it work.  This could be adopted and in the future if we enable explicit rotate then "auto rotate check" could be enabled when no number is passed, but if a number is passed, the image will be explicitly rotated the degrees specified and then run through face recognition at that angle.

**Comment by torzdf** [comment]
I'm not sure I follow your second point.

As to the first. the main issue with what your saying is that to detect all faces would mean rotating every single frame the full amount, which would lead to significant (and probably unacceptable) slowdown. It is still possible to use the --filter flag to filter faces with the image-rotator. I haven't used it myself, but if you are just looking to identify a particular face, this would seem to be the way to go, rather than looking to re-invent that functionality here. That way if the face you are looking for is not detected and another one is, then that face would be ignored, the rotation will be performed and it will scan again for the filtered face.

Ultimately I needed to make a trade-off between speed and effectiveness.  The best way to guarantee not to pick out a face you don't want is to mask the video prior to extracting faces.

That being said, I could change the command line flag to take options:
- off: default. Don't perform rotations (current faceswap behaviour)
- on : only rotate the image if a face isn't found (current rotator behaviour)
- force: rotate every frame all iterations (would capture all faces, and probably a lot more false positives, and would slow down the extract significantly)
- only-partial: Don't attempt to extract from upright frame, only perform rotations, stop when face is found
- only-full: Don't attempt to extract from upright frame, only perform rotations, but perform them all (again slow).

To be honest, this seems like overkill to me (I certainly would only ever use the first option).

I'm interested to see what other people's use case would be though, before embarking on a code overhaul. At the minimum, though, I could change the flag to accept arguments (just on or off, default off), so I could easily add options down the line.

**Comment by babilio** [comment]
I think you are right for now @torzdf. The problem right now is that there is nothing that deletes the alignments after extraction so if you do a run through with off, delete bad faces, run again with forced angle, it will have the bad entry in the alignments. So if you tried using skip extracted then it would get skipped anyways.

**Comment by torzdf** [comment]
@bryanlyon 

>  Obviously that isn't a huge deal, but changing the function of a command line switch is something we should avoid.

I've taken this on board, and changed the command line switch to an option (currently accepting on and off) so that I can more easily amend in future if needs be.

**Comment by babilio** [comment]
Yeah I think that's a good solution. Thanks for putting messages only on verbose mode too btw! Also retested this for performance by itself and convert is working great. Barely drop in images per second, even with images upside down.

**Comment by torzdf** [comment]
That's good to hear. 

In my opinion this is ready for merge if enough people are interested. I know @ppmdo and @Deadfishzzway showed interest in Issues. Tagging in case they want to test/feedback.

**Comment by torzdf** [comment]
@babilio How are your results, by the way?

@iperov mentioned:

> have you example where your fix can detect more faces without flip-glitches by aligner?

All my converts have come out fine, and I'm not sure what this potential glitch is, but it would be good to know whether you have run into any unexpected issues on the final output, or whether it's as expected.

**Comment by babilio** [comment]
They were good I had a long video that had even upside down faces. Like 1500 out of 8000 that got picked by the rotation only. They got converted back nicely too. This is great, I have always struggled with the awkward angles and how to spot faces and bulk rotate before faceswapping, this is so much more convinient

**Comment by ppmdo** [comment]
@torzdf Will test this asap and report.

**Comment by iperov** [comment]
@torzdf 
`All my converts have come out fine, and I'm not sure what this potential glitch is, but it would be good to know whether you have run into any unexpected issues on the final output, or whether it's as expected.`

so you dont know why your fix better, because you cannot provide any comparisons, and people have to test your fix ?

**Comment by babilio** [comment]
@iperov this fix is not meant to catch glitches on this extraction, it's to be able to detect sideway faces or upside down faces. Dlib expects eyes to be above nose. So if the person is like laying down on the side or face is upside down it won't get detected without this fix.

**Comment by iperov** [comment]
I just want to see comparison BEFORE and AFTER this subj.

**Comment by iperov** [comment]
for example
@torzdf extracted 100 faces, then he added rotations and extracted 105 faces - I want to see proofs

**Comment by babilio** [comment]
So the results I reported is not enough? Or are you saying that contributors should submit proof with their pull requests?

**Comment by babilio** [comment]
Yeah this I agree, this is a good practice actually

**Comment by Clorr** [comment]
A proof would mean the committer has tested things in various conditions and shows it works in all the cases. This can be expected when you have some critical project with hard consequences.

Here, @torzdf is proposing an improvment, that is only an option. Maybe in some cases it works maybe in other it doesn't. The PR enables people to have a look and say wehter this was useful to them or not or what behavior they encountered. I need absolutely no proof. If someone says "Oh it helped me in some cases", it is already good enough.

The only requirment for a PR is that the advantages overcome the drawbacks.

File: scripts/convert.py [hunk 1]
```diff
@@ -1,235 +1,241 @@
...
-            if is_have_alignments:
-                if self.have_face(filename):
-                    faces = self.get_faces_alignments(filename, image)
-                else:
-                    print ('no alignment found for {}, skipping'.format(os.path.basename(filename)))
-                    continue
-            else:
-                faces = self.get_faces(image)
-            yield filename, image, faces
+import cv2
```
Comment (Clorr) [inline scripts/convert.py]: There is new line problem in convert.py the diff shows all lines as changed while you just modified some. Can you plz check?

Comment (torzdf) [reply scripts/convert.py]: @Clorr Yeah, really annoyingly I merged with upstream master just before @oatssss fixed his newline issue, so the issue in that code pulled into my repo. Now my Git says that I'm in sync with upstream, although I'm clearly not :/

This was late last night, so I went to bed. Going to try to fix this morning.

I would ask that people refrain from testing until I get another commit in. I will update

Comment (Clorr) [reply scripts/convert.py]: No problem, this happens ;-)

**Comment by torzdf** [comment]
@Clorr should be fixed, so this should be good for pulling (I ran a quick test and all is fine and up to date).

@iperov I totally get where you are coming from, and if this was a more substantial change I would provide proof. However, I made this change mostly for myself. I have a deadline to hit so my dev box is currently running 24/7 performing converts (with the image rotator on as it happens). and I only have one usable gfx card, which is currently being utilized. Unfortunately the frames that are being generated cannot be shared in the public domain. 

I knew other people wanted this feature, so I raised this as a pull request knowing/hoping that other people would test it and confirm the results.

In an ideal world I would have provided an output, and if I commit a more substantial change in future, I will.

Cheers

**Comment by torzdf** [comment]
bloody hell, @Clorr. That was quick!

Thanks!

**Pull request merged**